### PR TITLE
[NVIDIA GPU] Assign a fixed index for cached activation

### PR DESCRIPTION
**Pull request by Tixxx**
gpu_windowed_einsum_handler pass has been re-using the empty buffer of the transformed while loop. This buffer is given by the spmd dot_handler pass. The shape of the buffer has changed from the allgathered shape of the sharded operand to the output shape of the dot which leads to a shape incompatibility error. To make the gpu handler completely safe, we will make a new element in the tuple to host the cached activation with the desired shape.
The slice index of where to write the slice into the full buffer also changes based on whether it's contracting or non-contracting dim is sharded. With the new element, we will need to determine the slice index ourselves in the handler pass.

### CODE REVIEW

**Comment by Tixxx** [comment]
@frgossen  can you take a look at this? This unblocks benchmarking from our side.

File: xla/tests/collective_ops_e2e_test.cc [hunk 1]
```diff
@@ -659,6 +659,7 @@ class CollectiveOpsTestE2EWindowedNonWindowed : public CollectiveOpsTestE2E {
     auto opts = GetDebugOptionsForTest();
     opts.set_xla_gpu_threshold_for_windowed_einsum_mib(0);
     opts.set_xla_gpu_multi_streamed_windowed_einsum(true);
+    opts.set_xla_gpu_mlir_emitter_level(0);
```
Comment (frgossen) [inline xla/tests/collective_ops_e2e_test.cc:662]: Why is this needed? Do you run into an issue with the MLIR emitters?

File: xla/service/gpu/gpu_windowed_einsum_handler_test.cc [hunk 2]
```diff
@@ -838,5 +837,82 @@ ENTRY main.9_spmd {
 )");
 }
 
+TEST_F(GpuWindowedEinsumHanlderTest,
+       AgLoopsMultipleConsumersAreChainedWithShardedContratingDim) {
+  constexpr absl::string_view kHloString = R"(
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler_test.cc:842]: Can you simplify the test case? From the PR description it reads like it is possible to reproduce this with a smaller example.

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 3]
```diff
@@ -434,24 +488,76 @@ absl::Status ProcessWindowedEinsumLoopForActivationCaching(
...
           while_body->AddInstruction(HloInstruction::CreateDynamicUpdateSlice(
               full_buffer_output_gte->shape(), full_buffer_output_gte,
-              cp_output,
+              first_cp_output,
               {dus_boundary_constant, slice_indices, dus_boundary_constant}));
     }
+
+    // If we have a DS(parameter_index1), then operands are sharded along the
+    // contracting dim. Slice indices will be the contracting dim's slices.
+    HloInstruction* slice_index;
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:500]: You are matching a very specific pattern here. Are you sure this is the right thing to do to fix this bug or does the issue occur for other patterns, too?

**Review by frgossen — CHANGES_REQUESTED** [review]

File: xla/tests/collective_ops_e2e_test.cc (hunk 1 above)
Comment (Tixxx) [reply xla/tests/collective_ops_e2e_test.cc:662]: i can remove this now. There was an issue with the mlir emitter level 1. We had a discussion on gchat, this has been changed to level 0 by default.

File: xla/service/gpu/gpu_windowed_einsum_handler_test.cc (hunk 2 above)
Comment (Tixxx) [reply xla/service/gpu/gpu_windowed_einsum_handler_test.cc:842]: The scenario that this breaks is specific, we need this pattern to trigger the issue:

1. an already rewritten allgather+gemm loop(with either contracting or noncontracting dim being sharded)
2. another allgather that shares the same operand with the above loop
3. another gemm that consumes the above allgather
That's why the unit test looks complex but it actually is a minimum case to repro. We had an unit test before with noncontracting dim sharded, this newly added one is with contracting dim sharded.

File: xla/service/gpu/gpu_windowed_einsum_handler.cc (hunk 3 above)
Comment (Tixxx) [reply xla/service/gpu/gpu_windowed_einsum_handler.cc:500]: Yea it's a good point. I simplified it a bit and made it more generic. We still need to be specific when we look for the slice index when we update the final output buffer when the loop is unrolled. I added more comment to explain why we need to be specific in that case.

**Review by frgossen — APPROVED** [review]

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 4]
```diff
@@ -434,24 +488,70 @@ absl::Status ProcessWindowedEinsumLoopForActivationCaching(
...
+    // the contracting dim. Slice indices will be the contracting dim's slices.
+    HloInstruction* slice_index;
+    HloInstruction* ds_index_constant;
+    HloInstruction* remainder;
+    HloInstruction* ds_param;
+    HloInstruction* ds_operand2;
+    HloInstruction* ds_operand3;
+    // There will be 2 dynamic-slices for unrolled loops, match for each one to
+    // get the slice index which will be used to write the corresponding
+    // received shard into cached activation buffer. For unrolled loops, we need
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:508]: Can you check this logic again? ds_operand2/3 are unused variables.

**Review by frgossen — CHANGES_REQUESTED** [review]

File: xla/service/gpu/gpu_windowed_einsum_handler.cc (hunk 4 above)
Comment (Tixxx) [reply xla/service/gpu/gpu_windowed_einsum_handler.cc:506]: ah forgot to remove them after cleanup. removed.

**Review by frgossen — APPROVED** [review]
